src/clj/fluree/db/reasoner/owl_datalog.cljc

```diff
@@ -0,0 +1,850 @@
+                            " with error: " (ex-message e))
+                       false))))
+
+(defn get-value
+  [val]
+  (if (map? val)
+    (or (:value val)
+        (get val "@value"))
+    val))
+
+(defn get-first-value
+  [node key]
+  (let [v (get-first node key)]
+    (get-value v)))
+
+(defn get-types
+  [node]
+  (or (:type node)
+      (get node "@type")))
+
+(defn get-all-ids
+  "Returns all @id values from either an ordered list or a set of objects.
+  Filters out any scalar (@value) values and blank nodes."
+  [vals]
+  (try*
+    (->> (util/sequential vals)
+         unwrap-list
+         (into [] (comp
+                    (map get-id)
+                    (remove nil?)
+                    (remove blank-node-id?))))
+    (catch* e
+            (throw (ex-info (str "Unexpected error extracting all @id values from owl statement:" vals)
+                            {:status 500
+                             :error  :db/unexpected-error}
+                            e)))))
+
+(defn equiv-class-type
+  [equiv-class-statement]
+  (let [statement-id (get-id equiv-class-statement)]
+    (cond (some #(= % const/iri-owl:Restriction) (get-types equiv-class-statement))
+          (cond
+            (contains? equiv-class-statement const/iri-owl:hasValue)
+            :has-value
+
+            (contains? equiv-class-statement const/iri-owl:someValuesFrom)
+            :some-values
+
+            (contains? equiv-class-statement const/iri-owl:allValuesFrom)
+            :all-values
+
+            (contains? equiv-class-statement const/iri-owl:maxCardinality)
+            :max-cardinality
+
+            (contains? equiv-class-statement const/iri-owl:maxQualifiedCardinality)
+            :max-qual-cardinality
+
+            (contains? equiv-class-statement const/iri-owl:qualifiedCardinality)
+            :qual-cardinality
+
+            :else
+            (do
+              (log/warn "Unsupported owl:Restriction" equiv-class-statement)
+              nil))
+
+          (contains? equiv-class-statement const/iri-owl:oneOf)
+          :one-of
+
+          (contains? equiv-class-statement const/iri-owl:intersectionOf)
+          :intersection-of
+
+          (contains? equiv-class-statement const/iri-owl:unionOf)
+          :union-of
+
+          statement-id
+          (if (blank-node-id? statement-id)
+            :blank-nodes
+            :classes)
+
+          :else nil)))
+
+(defmulti to-datalog (fn [rule-type inserts owl-statement all-rules]
+                       rule-type))
+
+(defmethod to-datalog ::eq-sym
+  [_ inserts owl-statement all-rules]
+  ;; note any owl:sameAs are just inserts into the current db
+  ;; the owl:sameAs rule is a base rule for any existing owl:sameAs
+  ;; that might already exist in the current db
+  (let [id      (get-id owl-statement)
+        sa-ids  (-> owl-statement
+                    (get const/iri-owl:sameAs)
+                    get-all-ids)
+        rule-id (str const/iri-owl:sameAs "(" id ")")
```

Comment:
This process of creating a rule id of the same format by concatenating strings is repeated a bunch of times below. I think a function to create rule ids would be useful.